File: ui/src/reconcileSpec.ts
/**
 * Reconciliation specs — **ambient CI for workbooks** (productivity loops, PR-B).
 *
 * A checked-in `.workbench/reconcile/<name>.toml` maps workbook cells and ranges
 * to callables in the workspace's *own* code. Approve it once and a save of the
 * workbook re-runs the reconciliation gate with the values those callables
 * produce: nobody fires a check, and seconds later the chip flips.
 *
 * ## Why this file, and why it holds almost no markup
 *
 * This is the tool's **descriptor** module — the one `tools.ts` imports, so it
 * is on the eager launch path. It therefore holds only what the shell needs
 * before the panel is ever opened: the store, the command, and the status chip.
 * The panel *body* — the rows, the approval dialog copy and their stylesheet —
 * lives in `panels/SpecPanel.tsx` behind a **dynamic `import()`** (`React.lazy`
 * plus a warm on idle), so none of it enters the entry chunk (`e2e/perf/
 * bundle.spec.ts` asserts what is *inside* it, not only what it weighs). That is
 * also why there is no JSX here: the four elements it does need are built with
 * `createElement`, which costs nothing.
 *
 * ## The trust surface, in the client's words
 *
 * Nothing here can make a spec run. Approving is a REST call that echoes the
 * digest the row was **rendered with**, so a spec whose bytes moved between the
 * render and the click is refused (409) rather than approved on the strength of
 * what was on screen. The panel renders `approval.covered` verbatim, because the
 * sentence the copy uses — *this spec, running exactly this code, on this
 * machine, until either changes* — is only worth saying if the list behind it is
 * one a person can read.
 *
 * ## No singleton
 *
 * Specs live in a map keyed by `name`, never a `store.activeSpec`: a workspace
 * routinely proves several workbooks, and the panel lists all of them. State
 * lives in a second zustand instance **in this module**, read by nothing outside
 * this capability and its panel — the `validation.ts` / `usage.ts` precedent, and
 * exactly the condition CLAUDE.md puts on a store that is not a slice of
 * `store.ts`.
 */

import type { DockviewApi, IDockviewPanelProps } from "dockview";
import { createElement, lazy, Suspense, type ReactElement } from "react";
import { create } from "zustand";

import * as api from "./api";
import { openPanel } from "./dock";
import type { ToolCommand, WorkbenchTool } from "./registry";
import type { SpecOutcome, SpecState, SpecStatus, WorkspaceEvent } from "./types";
import { ReconnectingSocket } from "./ws";

/** Stable contract: a saved layout references this panel by it (`docs/tools.md`). */
export const TOOL_ID = "specs";

/** Who the approval is recorded as. The server stamps the timestamp and the
 * digest; this is the only field the client supplies, and it is deliberately a
 * constant rather than a text box — the machine's user is the only person who
 * can be at this keyboard, and a free-text approver would be a field to forge. */
const APPROVER = "you";

// ---- the semantic mapping (no new colour) -----------------------------------

/** How one spec renders: a DESIGN.md token for its colour, the matching tint,
 * and the word that goes with it — colour is never the only signal (§7). */
export interface SpecVisual {
  token: string;
  bg: string;
  label: string;
}

/** A spec's *status* — whether it may run at all. Straight onto the existing
 * semantic ramp; nothing new is invented. */
const STATUS_VISUAL: Record<SpecStatus, SpecVisual> = {
  approved: { token: "--success", bg: "--success-bg", label: "Approved" },
  // Not red: nobody has said yes yet, which is the resting state of a spec
  // somebody checked in, not a failure.
  unapproved: { token: "--info", bg: "--info-bg", label: "Not approved" },
  // The code moved under an approval. This one has to *stop* a reader, so it
  // takes the warn ramp rather than the "could not judge" grey.
  stale: { token: "--warn", bg: "--warn-bg", label: "Needs re-approval" },
  invalid: { token: "--error", bg: "--error-bg", label: "Unreadable" },
};

/** A *run's* verdict. `blocked` shares the neutral grey with `skipped`: neither
 * is a disagreement about the numbers, both are "did not / could not judge". */
const OUTCOME_VISUAL: Record<SpecOutcome, SpecVisual> = {
  pass: { token: "--success", bg: "--success-bg", label: "Pass" },
  warn: { token: "--warn", bg: "--warn-bg", label: "Warn" },
  fail: { token: "--error", bg: "--error-bg", label: "Fail" },
  blocked: { token: "--agent-idle", bg: "--agent-idle-bg", label: "Blocked" },
  skipped: { token: "--agent-idle", bg: "--agent-idle-bg", label: "Skipped" },
};

export const statusVisual = (status: SpecStatus): SpecVisual => STATUS_VISUAL[status];
export const outcomeVisual = (outcome: SpecOutcome): SpecVisual => OUTCOME_VISUAL[outcome];

// ---- ordering and the reading -----------------------------------------------

/** Worst first, then by name. The order the panel lists them in, so the row
 * that needs a person is the row they see without scrolling. */
const STATUS_RANK: Record<SpecStatus, number> = {
  invalid: 0,
  stale: 1,
  unapproved: 2,
  approved: 3,
};

export function orderSpecs(specs: readonly SpecState[]): SpecState[] {
  return [...specs].sort(
    (a, b) => STATUS_RANK[a.status] - STATUS_RANK[b.status] || a.name.localeCompare(b.name),
  );
}

/**
 * Does this spec want a person right now?
 *
 * Two ways, and they are different questions with the same answer. The *spec*
 * needs one when its code moved or its file will not parse — nothing is running,
 * and a loop that went quiet is indistinguishable from one that broke. The *run*
 * needs one when the last verdict was a fail or a refusal.
 *
 * A spec nobody has approved yet is deliberately **not** counted: it is not a
 * problem, it is a thing somebody checked in and has not armed. Counting it
 * would put a number on the bar of every workspace that ever contained a spec.
 */
export function needsAttention(spec: SpecState): boolean {
  if (spec.status === "stale" || spec.status === "invalid") return true;
  const outcome = spec.last_run?.outcome;
  return outcome === "fail" || outcome === "blocked";
}

/** How many specs need a person — the status-bar reading, which hides at zero
 * (DESIGN §6.7). A quiet bar means nothing needs you. */
export function attentionCount(specs: readonly SpecState[]): number {
  return specs.filter(needsAttention).length;
}

/** One line for the chip's tooltip. Names the specs rather than only counting
 * them: a number with no noun is a thing you have to click to understand. */
export function attentionTitle(specs: readonly SpecState[]): string {
  const wanting = orderSpecs(specs.filter(needsAttention));
  if (wanting.length === 0) return "Every reconciliation spec is passing";
  const named = wanting.slice(0, 3).map((spec) => `${spec.name} (${summaryOf(spec)})`);
  const more = wanting.length > 3 ? `, and ${String(wanting.length - 3)} more` : "";
  return `${named.join(", ")}${more}`;
}

/** The shortest true thing about a spec — the word a row's pill carries. */
export function summaryOf(spec: SpecState): string {
  if (spec.status !== "approved") return STATUS_VISUAL[spec.status].label.toLowerCase();
  const outcome = spec.last_run?.outcome;
  return outcome === undefined ? "not run yet" : OUTCOME_VISUAL[outcome].label.toLowerCase();
}

// ---- the store ---------------------------------------------------------------

interface SpecStore {
  /** Held specs, keyed by `name`. Empty is a real answer (a workspace with no
   * specs) and the common one — never a singleton. */
  specs: Record<string, SpecState>;
  /** The server's sentence about the *set* — what an empty list means, which is
   * a thing blankness cannot say (AXI shape 2, applied to a person). */
  detail: string;
  /** `false` until the first `refresh()` settles either way. Before it flips, an
   * absent name is *unknown*, not gone — the `validation.ts` posture. */
  hydrated: boolean;
  /** The last call that failed, as one sentence. Cleared by the next attempt. */
  error: string | null;
  init: () => void;
  refresh: () => Promise<void>;
  /** The one-time trust decision, echoing the digest this row was rendered with.
   * A 409 (the spec moved under the dialog) surfaces as `error` and re-reads,
   * never as a decision that silently did not land. */
  approve: (name: string) => Promise<void>;
  revoke: (name: string) => Promise<void>;
  /** Run one now. The loop's point is that nobody has to, so this is the
   * affordance for the first run after an approval and for impatience. */
  run: (name: string) => Promise<void>;
}

let started = false;

/** Reset the module-level "already subscribed" latch and the map. Tests only. */
export function resetSpecStoreForTests(): void {
  started = false;
  useSpecStore.setState({ specs: {}, detail: "", hydrated: false, error: null });
}

function message(error: unknown): string {
  return error instanceof Error ? error.message : String(error);
}

export const useSpecStore = create<SpecStore>((set, get) => ({
  specs: {},
  detail: "",
  hydrated: false,
  error: null,

  init: () => {
    if (started) return;
    started = true;
    void get().refresh();
    new ReconnectingSocket("/ws/events", {
      onMessage: (data) => {
        const event = data as WorkspaceEvent;
        if (event.type !== "reconcile_spec") return;
        // This frame is the whole product claim arriving: a run nobody fired,
        // pushed to a window that never asked for it.
        set((state) => ({ specs: { ...state.specs, [event.state.name]: event.state } }));
      },
      // Every reconnect re-reads: a run that happened while the socket was down
      // is gone from the live stream, and the load path is the only way back.
      onOpen: () => {
        void get().refresh();
      },
    });
  },

  refresh: async () => {
    try {
      const { specs, detail } = await api.getSpecs();
      set({
        specs: Object.fromEntries(specs.map((spec) => [spec.name, spec])),
        detail,
        hydrated: true,
        error: null,
      });
    } catch (error) {
      // A server that cannot answer leaves the last map standing — the honest
      // reading, and the usage/activity posture. Still hydrated: the store
      // *answered*, so an absent name is now genuinely absent.
      set({ hydrated: true, error: message(error) });
    }
  },

  approve: async (name) => {
    const spec = get().specs[name];
    if (spec === undefined) return;
    set({ error: null });
    try {
      const updated = await api.approveSpec(name, {
        approver: APPROVER,
        digest: spec.digest,
      });
      set((state) => ({ specs: { ...state.specs, [updated.name]: updated } }));
    } catch (error) {
      // Whatever went wrong, the row on screen is now a row nobody should act
      // on again until it has been re-read — the 409 case *is* "your copy is
      // stale". Re-read first and set the message after, because `refresh`
      // clears the error on success and the user has to keep the sentence that
      // told them why their click did nothing.
      await get().refresh();
      set({ error: message(error) });
    }
  },

  revoke: async (name) => {
    set({ error: null });
    try {
      const updated = await api.revokeSpec(name);
      set((state) => ({ specs: { ...state.specs, [updated.name]: updated } }));
    } catch (error) {
      set({ error: message(error) });
    }
  },

  run: async (name) => {
    set({ error: null });
    try {
      // The report also arrives on the socket as a whole `SpecState`; re-reading
      // is what turns it into the row, and it is idempotent.
      await api.runSpec(name);
      await get().refresh();
    } catch (error) {
      set({ error: message(error) });
    }
  },
}));

// ---- lifecycle ---------------------------------------------------------------

/** The panel body's chunk. `React.lazy` caches the import, so warming it means
 * the panel mounts without suspending. Idempotent and cheap once warm. */
async function loadPanel(): Promise<void> {
  try {
    await import("./panels/SpecPanel");
  } catch {
    // The panel opens on the Suspense fallback instead; not worth blocking on.
  }
}

function openSpecs(): void {
  useSpecStore.getState().init();
  void useSpecStore.getState().refresh();
  void loadPanel().then(() => {
    openPanel(TOOL_ID);
  });
}

function warmPanel(): void {
  const warm = (): void => void loadPanel();
  const idle = (globalThis as { requestIdleCallback?: (cb: () => void) => number })
    .requestIdleCallback;
  if (idle === undefined) setTimeout(warm, 1_000);
  else idle(warm);
}

function onDockReady(dock: DockviewApi | null): void {
  if (dock === null) return;
  useSpecStore.getState().init();
  warmPanel();
}

/** The workspace moved: a different project has different specs, and different
 * approvals. Re-read rather than carry — a row about `dispatch` in the folder
 * you left is not a row about `dispatch` in the one you opened. */
function onWorkspaceChanged(): void {
  useSpecStore.setState({ specs: {}, hydrated: false, error: null });
  void useSpecStore.getState().refresh();
}

// ---- the panel wrapper and the chip (no JSX: see the module note) ------------

const SpecPanelLazy = lazy(() => import("./panels/SpecPanel"));

function SpecsPanel(props: IDockviewPanelProps): ReactElement {
  return createElement(
    Suspense,
    { fallback: createElement("div", { className: "wb-spec-loading u-label" }, "Loading…") },
    createElement(SpecPanelLazy, props),
  );
}

/** The tab glyph: a workbook grid with a tick — the two ideas this capability
 * joins, and deliberately not the Review panel's shield. */
function SpecsIcon(): ReactElement {
  return createElement(
    "svg",
    {
      width: 14,
      height: 14,
      viewBox: "0 0 16 16",
      fill: "none",
      stroke: "currentColor",
      strokeWidth: 1.2,
      strokeLinecap: "round" as const,
      strokeLinejoin: "round" as const,
      "aria-hidden": true,
    },
    createElement("rect", { key: "grid", x: 2, y: 2.5, width: 8, height: 11, rx: 1 }),
    createElement("path", { key: "rows", d: "M2 6h8M6 2.5v11" }),
    createElement("path", { key: "tick", d: "M9.5 11l2 2 3.2-4" }),
  );
}

/**
 * The status chip. Hidden at zero, per the quiet-bar doctrine — and that is the
 * whole point of this capability: a bar with nothing on it is a bar saying every
 * workbook still agrees with the code.
 */
export function SpecStatusChip(): ReactElement | null {
  const specs = useSpecStore((state) => state.specs);
  const rows = Object.values(specs);
  const count = attentionCount(rows);
  if (count === 0) return null;
  return createElement(
    "button",
    {
      type: "button",
      className: "wb-status-chip",
      title: attentionTitle(rows),
      onClick: openSpecs,
    },
    createElement("span", { key: "glyph", "aria-hidden": true }, "⌗"),
    createElement("span", { key: "count", className: "u-tabular" }, String(count)),
    createElement("span", { key: "noun" }, count === 1 ? "spec" : "specs"),
  );
}

// ---- registration -------------------------------------------------------------

const commands: readonly ToolCommand[] = [
  {
    id: "specs.open",
    title: "Reconciliation specs…",
    detail: () => "approve a spec once; every save of its workbook re-runs the gate",
    run: openSpecs,
  },
];

export const specsTool: WorkbenchTool = {
  id: TOOL_ID,
  title: "Specs",
  icon: SpecsIcon,
  panel: {
    component: SpecsPanel,
    // Right, beside Review: the two are the same idea at two distances — this
    // one says which proofs are armed, that one says what one proof found.
    defaultLocation: { area: "right", size: 340 },
    openByDefault: false,
    // SINGULAR, and here is the reason the pane rules ask for. A pane of this
    // tool points at no resource: it is the *list* of the workspace's specs, of
    // which there is exactly one, and a second pane would be a second view of
    // the same rows each able to approve over the other's decision. The plural
    // surface this capability really has is the Review pane — one per
    // `ValidationResult`, which is what a run produces.
    singleton: true,
  },
  commands,
  // No Alt chord by default (the Review/Workspaces precedent): a registered
  // chord is one the user's own `shortcuts.md` can never have, and approving a
  // spec is not a reflex. It is bindable from `shortcuts.md` the day it lands.
  statusContributions: [{ region: "right", component: SpecStatusChip }],
  onDockReady,
  onWorkspaceChanged,
};

File: server/src/workbench_server/models/reconcile_spec.py
"""Spec-from-code: the typed shapes behind ``.workbench/reconcile/<name>.toml``.

A :class:`~workbench_server.models.reconciliation.ReconciliationSpec` takes its
``expected`` values **as data**, on purpose — ``models/reconciliation.py`` is
explicit that executing user code out of a JSON body is what the never-execute
doctrine forbids. That decision stands and nothing here weakens it. What it left
open is that *somebody* has to produce the data every time, which is why the gate
was fired by hand and therefore fired rarely.

This module is the other end: a **checked-in file** names the analyst's own
callable, the server owns the argv, and a **one-time content-hash approval** is
what turns a file in a folder into something that may run.

Three properties are in the schema rather than in a comment, because each is a
promise a later edit could quietly break:

* **Nothing in here can express a command.** A :class:`SpecCheck` names a
  ``module:function`` *within the workspace* and nothing else — no argv, no cwd,
  no interpreter. The argv is fixed and server-owned
  (``services/reconcile_spec.py``), and the spec travels to it on **stdin**, so
  ``models/gates.py``'s rule ("there is no field anywhere in this module through
  which a JSON body can reach an argv, a cwd or a path") holds here too. The
  posture that *is* wider than the gates' — this runs in the live workspace root,
  which ``ToolchainGateCheck`` refuses — is paid for by the approval below.

* **The approval is keyed to the code, not merely to the file that names it.**
  A :class:`SpecApproval` carries a composite digest over a list of
  :class:`CoveredSource` — the spec's own bytes, the module each ``callable``
  resolves to, and the workspace-local import closure a previous run *actually*
  used. Hashing the ``.toml`` alone would be a hole with a name: approve once,
  then rewrite the body of ``annual_revenue``, and the spec file — and therefore
  the approval — is untouched while the watcher keeps running whatever that
  function now contains, unattended, forever. An approval that authorises code it
  never hashed is not a trust prompt; it is a trust prompt's shadow.

* **What is covered is a list a person can read.** :attr:`SpecApproval.covered`
  is rendered by the panel verbatim, so "this spec, running exactly this code, on
  this machine, until either changes" is a claim with a receipt rather than one
  taken on faith.

Only :class:`SpecStates`, :class:`SpecState`, :class:`SpecApprovalRequest` and
:class:`SpecEvent` are wire bodies of their own. :class:`SpecEntryRequest` /
:class:`SpecEntryResult` are the **stdin/stdout envelope** of the subprocess and
never reach a browser; they are typed all the same, because the boundary they
cross is the one that matters most.
"""

from __future__ import annotations

from datetime import datetime
from pathlib import Path
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, model_validator

from workbench_server.models.reconciliation import Tolerance

#: Where specs live, relative to the workspace root. A *folder* rather than one
#: file because a workspace routinely proves more than one workbook, and because
#: the approval is per spec: one bad edit revokes one spec, not the set.
SPEC_DIR = Path(".workbench") / "reconcile"

#: Suffix a spec file must carry. TOML because it is a file a person edits.
SPEC_SUFFIX = ".toml"

#: Ceiling on one spec file, in bytes. A spec is a page of declarations; anything
#: past this is not a spec that got long, it is a file that wandered into the
#: folder. Refused with the size, never silently half-read.
MAX_SPEC_BYTES = 256 * 1024

#: Ceiling on how many specs one workspace's folder contributes. Bounded because
#: every one of them is re-digested on every listing.
MAX_SPECS = 64

#: Ceiling on the checks in one spec. Each check is one callable in one
#: subprocess run, so this bounds the work a single save can buy.
MAX_CHECKS_PER_SPEC = 64

#: Ceiling on the ``(timestamp, value)`` pairs one ``range`` check may return.
#: A full hourly year is 8,760; this leaves room for a decade of them and still
#: bounds a callable that decided to return its whole dataframe.
MAX_PAIRS_PER_CHECK = 100_000

#: Ceiling on the bytes of the subprocess's **stdout** — the JSON envelope. The
#: envelope is bounded by the two caps above; this is the backstop for a child
#: that writes to the real stdout anyway. Chatty output goes to *stderr* (the
#: entry module redirects the callable's own ``print``), where it is bounded by
#: the gates' ``_BoundedCapture``.
MAX_SPEC_STDOUT_BYTES = 8 * 1024 * 1024

#: Bytes of the child's stderr kept for the evidence line. Head + ring tail, the
#: gates' window, for the same reason: a callable that prints a 500 MB dataframe
#: costs the window, not the memory.
MAX_SPEC_LOG_BYTES = 8_192

#: Env knobs, quoted in every refusal that they would fix.
SETTING_TIMEOUT = "WORKBENCH_RECONCILE_TIMEOUT_S"
SETTING_FAKE = "WORKBENCH_RECONCILE_FAKE"


# ---- the file ----------------------------------------------------------------


class SpecCheck(BaseModel):
    """One check: a workbook address, and the callable that produces its expected
    value.

    Exactly one of :attr:`cell` and :attr:`range` — a check that addressed both
    would have two answers about what shape the callable must return, and a check
    that addressed neither is a typo the reader should hear about at load time.
    """

    model_config = ConfigDict(populate_by_name=True)

    #: A1 address, sheet-qualified (``Summary!D14``) or bare (``D14`` → active
    #: sheet). The callable must return a scalar.
    cell: str | None = None
    #: A two-column A1 range, ``Hours!A2:B8761`` — timestamp column then value
    #: column. The callable must return an iterable of
    #: ``(naive local datetime, float)`` pairs, which is exactly ``TimeIndexSpec``'s
    #: vocabulary, so a spec compiles into a ``ReconciliationSpec`` and nothing
    #: downstream of the check changes.
    range: str | None = None
    #: ``module:function`` **within the workspace**. Field name is ``call``
    #: because ``callable`` is a builtin; the alias is what the TOML and the wire
    #: both use, so a person never types the compromise.
    call: str = Field(alias="callable")
    #: Unit of the value the *callable* returns.
    unit: str = ""
    #: Declared unit of the *workbook* side. ``None`` means "same as
    #: :attr:`unit`". A differing, compatible pair is a **named** conversion in
    #: the evidence — the x1000 this domain hides bugs in is never silent.
    value_unit: str | None = None
    label: str | None = None
    #: Per-check override of the spec's ``default_tolerance``.
    tolerance: Tolerance | None = None

    @model_validator(mode="after")
    def _exactly_one_address(self) -> SpecCheck:
        if (self.cell is None) == (self.range is None):
            raise ValueError("a check names exactly one of `cell` or `range`")
        if ":" not in self.call:
            raise ValueError(f"`callable` must be module:function, got {self.call!r}")
        module, _, function = self.call.partition(":")
        if not module or not function:
            raise ValueError(f"`callable` must be module:function, got {self.call!r}")
        return self

    def address(self) -> str:
        """The address this check names, whichever field carries it."""
        return self.cell if self.cell is not None else str(self.range)


class ReconcileSpecFile(BaseModel):
    """One ``.workbench/reconcile/<name>.toml``, parsed.

    Deliberately *not* a wire body: it is read from disk, never posted. The
    browser sees :class:`SpecState`, which is this file plus what the server
    knows about it.
    """

    model_config = ConfigDict(populate_by_name=True)

    #: Workspace-relative path to the ``.xlsx``, jailed against the root by the
    #: service before anything opens it.
    workbook: str
    #: IANA zone for ``range`` checks, e.g. ``Europe/Oslo``. Required when any
    #: check names a range: a time-indexed comparison without a zone cannot tell
    #: a fall-back day's two 02:00s apart, which is the whole reason the aligner
    #: exists.
    timezone: str | None = None
    default_tolerance: Tolerance
    #: ``[[check]]`` in the TOML — singular there because that is what a person
    #: writing one at a time types.
    checks: list[SpecCheck] = Field(
        default_factory=list, alias="check", max_length=MAX_CHECKS_PER_SPEC
    )

    @model_validator(mode="after")
    def _zone_when_time_indexed(self) -> ReconcileSpecFile:
        if any(check.range is not None for check in self.checks) and self.timezone is None:
            raise ValueError("a `range` check needs a top-level `timezone`")
        return self


# ---- what an approval covers -------------------------------------------------

#: How a covered file entered the approval. ``spec`` is the ``.toml`` itself,
#: ``callable`` is a module a ``callable`` entry resolved to at approval time,
#: and ``imported`` is a workspace-local module a *previous run* actually used
#: (the ``sys.modules`` closure — a fact, not a guess at the import graph).
CoveredOrigin = Literal["spec", "callable", "imported"]


class CoveredSource(BaseModel):
    """One file an approval stands for, and the bytes it stood for."""

    #: Workspace-relative, POSIX separators. Relative because an approval that
    #: moved with the folder would be an approval nobody re-made.
    path: str
    #: ``blake2b`` of the file's bytes, or :data:`ABSENT_DIGEST` when it is gone.
    digest: str
    origin: CoveredOrigin


#: The digest recorded for a covered path that cannot be read. A distinct value
#: rather than an omission: "the file is gone now" is a state that has to be
#: tellable apart from "nothing changed" (``services/gates.py::content_digest``,
#: same rule one level up).
ABSENT_DIGEST = "absent"


class SpecApproval(BaseModel):
    """The one-time trust decision, recorded.

    Stored in the machine's app-data dir and **never** in ``.workbench/`` — the
    reason ``RecentsStore`` spells out, sharpened by what this one authorises: a
    trust record inside the folder it authorises is a trust record an attacker
    can write.
    """

    #: Spec file stem, e.g. ``dispatch`` for ``dispatch.toml``.
    name: str
    #: The composite digest over :attr:`covered`, in the order that list carries.
    #: Re-computed before every run and compared; a mismatch is a refusal, never
    #: a silent re-run and never a silent skip.
    digest: str
    approver: str
    #: Server-minted, never taken from the caller.
    approved_at: datetime
    #: Every file this approval stands for, rendered verbatim by the panel.
    covered: list[CoveredSource] = Field(default_factory=list)


# ---- what a run produced -----------------------------------------------------

#: A spec's own verdict. ``blocked`` is the digest-mismatch state — the code
#: changed and nobody re-approved it — and it is deliberately its own word rather
#: than a ``skipped``: the difference between them is the difference between a
#: badge someone scrolls past and a badge that stops them.
#:
#: Note for the reader coming from the plan: ``CheckOutcome`` (``models/
#: validation.py``) does not carry ``"blocked"`` yet — that row is PR-A's, in
#: PR-A's files. Until it lands, a blocked *spec* still produces a ``fail``
#: evidence line carrying the same sentence, so the risk is ``high`` and nothing
#: is ever waved through; only the word on the pill is the compromise.
SpecOutcome = Literal["pass", "warn", "fail", "blocked", "skipped"]

#: Whether a spec may run right now, and why not when it may not.
SpecStatus = Literal["unapproved", "approved", "stale", "invalid"]


class SpecRunReport(BaseModel):
    """One whole run of one spec — what it judged and how long it took."""

    name: str
    outcome: SpecOutcome
    #: The ``ValidationResult`` this run produced, when it produced one. ``None``
    #: for a run that never reached the gate (unapproved, stale, invalid).
    validation_id: str | None = None
    #: What started it — the loop's whole claim is that ``watcher`` happens
    #: without anybody clicking, so the two are told apart on the record.
    trigger: Literal["watcher", "manual"] = "manual"
    ran_at: datetime
    duration_ms: int
    #: How many expected values the callables produced (scalars + pairs).
    values: int = 0
    #: One sentence a human reads. Never blank: an empty result says it is empty
    #: (AXI shape 2) and a refusal names the way to fix it (shape 3).
    detail: str


class SpecState(BaseModel):
    """One row of the panel: a spec, what it points at, and whether it may run."""

    name: str
    #: Workspace-relative path of the ``.toml``.
    path: str
    #: Workspace-relative path of the workbook it proves. Empty when the file
    #: could not be parsed at all.
    workbook: str = ""
    status: SpecStatus
    #: The composite digest **as it is right now**. This is the value the client
    #: echoes back on approve, so a spec that changed between the render and the
    #: click is refused rather than approved on the strength of what was on
    #: screen.
    digest: str
    checks: int = 0
    approval: SpecApproval | None = None
    #: The last run this server observed, or ``None``. In memory: a restart
    #: forgets it, and the approval — which is the part that matters — does not.
    last_run: SpecRunReport | None = None
    #: One sentence: why it cannot run, or what it is proving.
    detail: str


class SpecStates(BaseModel):
    """``GET /api/reconcile/specs`` — every spec in the workspace's folder.

    An empty list is a real and common answer (a workspace with no specs), and
    :attr:`detail` says so rather than leaving blankness to interpret.
    """

    specs: list[SpecState] = Field(default_factory=list)
    #: One sentence about the set — "no specs" names the folder to create.
    detail: str


class SpecApprovalRequest(BaseModel):
    """``POST /api/reconcile/specs/{name}/approve``.

    The caller echoes the digest it was shown. A digest that no longer matches is
    **409**, not a 200: approving a spec whose bytes moved under the dialog would
    approve something nobody read.
    """

    approver: str
    digest: str


class SpecEvent(BaseModel):
    """Broadcast on ``/ws/events`` whenever a spec's state changes.

    Carries the whole state, not a delta — the client holds its own map keyed by
    ``name`` and replaces the entry, exactly as it does for a ``ValidationEvent``.
    """

    type: Literal["reconcile_spec"] = "reconcile_spec"
    state: SpecState


# ---- the subprocess envelope -------------------------------------------------


class SpecEntryRequest(BaseModel):
    """What travels to the child on **stdin**. Never an argv.

    The child is started as a fixed, server-owned
    ``uv run python -m workbench_server.spec_entry`` with the workspace root as
    its cwd; this document is the only variable input, and every field in it is a
    *selection* the spec file already made.
    """

    #: ``module:function`` strings, in the spec's own order. The child imports
    #: each and calls it with no arguments.
    callables: list[str] = Field(default_factory=list, max_length=MAX_CHECKS_PER_SPEC)
    #: Ceiling on the pairs one callable may return, so a dataframe cannot
    #: become an unbounded stdout.
    max_pairs: int = MAX_PAIRS_PER_CHECK


class SpecValue(BaseModel):
    """What one callable answered."""

    #: The ``module:function`` this answers for, echoed so the parent never has
    #: to trust list order.
    call: str
    ok: bool
    #: A scalar answer, for a ``cell`` check.
    scalar: float | None = None
    #: ``(naive local ISO timestamp, value)`` pairs, for a ``range`` check.
    pairs: list[tuple[str, float]] = Field(default_factory=list)
    #: Set when the pair list was cut at ``max_pairs`` (AXI shape 1).
    total_pairs: int | None = None
    #: Why it failed, when it did. One line, already bounded by the child.
    error: str | None = None


class SpecEntryResult(BaseModel):
    """What comes back on the child's **stdout** — one compact JSON line.

    :attr:`modules` is a list of **paths only**, deliberately. The child names
    the files it imported; the *parent* hashes them, because a digest supplied by
    the process being trusted is not evidence of anything.
    """

    ok: bool
    values: list[SpecValue] = Field(default_factory=list)
    #: Absolute paths of every module in ``sys.modules`` whose ``__file__``
    #: resolves under the workspace root after the callables returned. A fact
    #: about what ran, not a static guess at an import graph that may be
    #: conditional or built with ``importlib``.
    modules: list[str] = Field(default_factory=list)
    #: Set when the child could not even start the work (a malformed request).
    error: str | None = None
